### Dinda Hana Zhafira <br> 22/494038/PA/21240

---
Pada eksplorasi data kali ini, terdapat 2 tahap yaitu data acquisition dan penarikan kesimpulan berdasarkan analisis menggunakan statistical method



# Data Acquisition
Data diperoleh dari website open-source kaggle. Dataset yang digunakan adalah diabetes dataset yang berasal dari The National Institute of Diabetes and Digestive and Kidney Diseases.

1. Berikan akses kepada google drive

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


2. Import modul untuk mengupload file dari local system

In [ ]:
from google.colab import files

files.upload()

Saving kaggle.json to kaggle.json


{'kaggle.json': b'{"username":"dindahanazhafira","key":"8e72831daa6c2e4a26e707f323e068f9"}'}

3. Pastikan `kaggle.json` sudah ada

In [ ]:
!ls -lha kaggle.json

-rw-r--r-- 1 root root 72 Apr  5 14:20 kaggle.json


4. Install kaggle API client

In [ ]:
!pip install -q kaggle

5. Salin file `kaggle.json` ke dalam direktori `~/.kaggle/`. File ini berisi kredensial API Kaggle yang diperlukan untuk mengunduh dataset dari Kaggle API.

In [ ]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/

6. Atur izin file `kaggle.json`

In [ ]:
!chmod 600 /root/.kaggle/kaggle.json

7. Cek directory sebelum mendownload datasets

In [ ]:
!pwd

/content


8. Menampilkan list dataset yang ada

In [ ]:
!kaggle datasets list

ref                                                         title                                           size  lastUpdated          downloadCount  voteCount  usabilityRating  
----------------------------------------------------------  ---------------------------------------------  -----  -------------------  -------------  ---------  ---------------  
sudarshan24byte/online-food-dataset                         Online Food Dataset                              3KB  2024-03-02 18:50:30          19175        388  0.9411765        
sahirmaharajj/electric-vehicle-population-size-2024         Electric Vehicle Population by Country (2024)  275KB  2024-03-30 19:16:06           1098         32  1.0              
mexwell/drug-consumption-classification                     💊 Drug Consumption Classification               56KB  2024-03-28 15:02:22           1585         34  1.0              
bhavikjikadara/student-study-performance                    Student Study Performance                    

9. Download dataset menggunakan API dari kaggle

In [ ]:
!kaggle datasets download -d akshaydattatraykhare/diabetes-dataset

  0% 0.00/8.91k [00:00<?, ?B/s]
100% 8.91k/8.91k [00:00<00:00, 17.8MB/s]


10. Ekstrak file zip yang telah didownload

In [ ]:
!unzip diabetes-dataset.zip

Archive:  diabetes-dataset.zip
  inflating: diabetes.csv            


Dataset yang digunakan adalah diabetes dataset dengan atribut:
- Pregnancies: menunjukkan jumlah kehamilan
- Glucose: menunjukkan level glukosa dalam darah
- BloodPressure: menunjukkan hasil pengukuran tekanan darah
- SkinThickness: menunjukkan ketebalan kulit
- Insulin: menunjukkan level insulin dalam darah
- BMI: menunjukkan body mass index
- DiabetesPedigreeFunction: menunjukkan persentasi kemungkinan terkena diabetes
- Age: menunjukkan umur
- Outcome: menunjukkan hasil akhir, 1 berarti yes dan 0 berarti no

# Statistical Method

### Import library `numpy` dan `pandas` untuk perhitungan statistik

In [ ]:
import pandas as pd
import numpy as np

----

### Baca file CSV `diabetes.csv`

In [ ]:
path = "diabetes.csv"
df = pd.read_csv(path)

-----

### Tampilkan 10 list teratas di dataset

In [ ]:
df.head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   
5            5      116             74              0        0  25.6   
6            3       78             50             32       88  31.0   
7           10      115              0              0        0  35.3   
8            2      197             70             45      543  30.5   
9            8      125             96              0        0   0.0   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  
5                     0.201   30        0  
6                     0.248   26        1  
7                     0.134   29        0  
8                     0.158   53        1  
9                     0.232   54        1

----

### Tampilkan 10 list terbawah

In [ ]:
df.tail(10)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
758            1      106             76              0        0  37.5   
759            6      190             92              0        0  35.5   
760            2       88             58             26       16  28.4   
761            9      170             74             31        0  44.0   
762            9       89             62              0        0  22.5   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
758                     0.197   26        0  
759                     0.278   66        1  
760                     0.766   22        0  
761                     0.403   43        1  
762                     0.142   33        0  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0

-----

### Menghitung jumlah berdasarkan outcome

In [ ]:
df['Outcome'].value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

- Dari data di atas, didapat bahwa orang yang positif diabetes (disimbolkan angka 1) sebanyak 268 orang dan lebih sedikit dibandingkan orang yang negatif diabetes yaitu 500 orang
- Data yang didapat adalah imbalanced data

### Tampilkan data type dari setiap column

In [ ]:
print(df.dtypes)

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


----

### Tampilkan perhitungan statistik untuk semua kolom baik numerik maupun non-numerik

In [ ]:
df.describe(include='all')

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

----

### Hitung mean untuk sebagai parameter untuk menarik kesimpulan

In [ ]:
df.groupby('Outcome')[['Pregnancies', 'Glucose', 'BloodPressure',	'SkinThickness', 'Insulin',	'BMI',	'DiabetesPedigreeFunction', 'Age']].mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

- Rata - rata `Glucose` dan `Insulin` secara signifikan lebih tinggi pada orang positif diabetes dibandingkan yang negatif. Hal ini menunjukkan bahwa seseorang dengan glukosa dan insulin yang tinggi memiliki kecenderungan untuk positif diabetes
- `BloodPressure` dan `SkinThickness` tidak menunjukkan perbedaan yang signifikan antara kedua kelompok.
- Rata - rata `BMI` (Body Mass Index) lebih tinggi pada kelompok yang memiliki diabetes (Outcome 1) dibandingkan dengan kelompok yang tidak memiliki diabetes (Outcome 0), menunjukkan korelasi antara obesitas atau overweight dengan diabetes tipe 2.
- Rata - rata `Diabetes Pedigree Function` lebih tinggi pada kelompok yang memiliki diabetes (Outcome 1) dibandingkan dengan kelompok yang tidak memiliki diabetes (Outcome 0), menunjukkan adanya faktor risiko genetik.
- Rata - rata `Age` lebih tinggi pada kelompok yang memiliki diabetes (Outcome 1) dibandingkan dengan kelompok yang tidak memiliki diabetes (Outcome 0), yang konsisten dengan fakta bahwa risiko diabetes meningkat dengan bertambahnya usia.

----

### Hitung median untuk sebagai parameter untuk menarik kesimpulan

In [ ]:
df.groupby('Outcome')[['Pregnancies', 'Glucose', 'BloodPressure',	'SkinThickness', 'Insulin',	'BMI',	'DiabetesPedigreeFunction', 'Age']].median()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin    BMI  \
Outcome                                                                       
0                2.0    107.0           70.0           21.0     39.0  30.05   
1                4.0    140.0           74.0           27.0      0.0  34.25   

         DiabetesPedigreeFunction   Age  
Outcome                                  
0                           0.336  27.0  
1                           0.449  36.0

- Pada `Pregnancies`, `Age`, `Glucose`, `SkinThickness`, `BloodPressure`, `BMI`, dan `DiabetesPedigreeFunction`, mediannya lebih tinggi pada kelompok yang memiliki diabetes (Outcome 1) dibandingkan dengan kelompok yang tidak memiliki diabetes (Outcome 0).
- Pada `Insulin`, median kadarnya dalam darah tidak dapat diukur pada kelompok yang memiliki diabetes (Outcome 1) karena semua nilai pada kelompok ini adalah 0, sedangkan median untuk kelompok yang tidak memiliki diabetes (Outcome 0) adalah 39.0.

---

### Hitung modus untuk sebagai parameter untuk menarik kesimpulan

In [ ]:
df.mode()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0          1.0       99           70.0            0.0      0.0  32.0   
1          NaN      100            NaN            NaN      NaN   NaN   

   DiabetesPedigreeFunction   Age  Outcome  
0                     0.254  22.0      0.0  
1                     0.258   NaN      NaN

----

### Hitung varians sebagai parameter untuk menarik kesimpulan

In [ ]:
df.groupby('Outcome')[['Pregnancies', 'Glucose', 'BloodPressure',	'SkinThickness', 'Insulin',	'BMI',	'DiabetesPedigreeFunction', 'Age']].var()

Pregnancies      Glucose  BloodPressure  SkinThickness       Insulin  \
Outcome                                                                         
0           9.103403   683.362325     326.274693     221.710525   9774.345427   
1          13.996870  1020.139457     461.897968     312.572195  19234.673319   

               BMI  DiabetesPedigreeFunction         Age  
Outcome                                                   
0        59.133870                  0.089452  136.134168  
1        52.750693                  0.138648  120.302588

- Kita dapat melihat bahwa `variansi` dalam banyak fitur yang diukur lebih besar pada kelompok yang memiliki diabetes (Outcome 1), menunjukkan tingkat heterogenitas yang lebih besar dalam data untuk kelompok ini. Hal ini konsisten dengan kompleksitas dan variasi dalam gejala dan komplikasi diabetes.

----

### Hitung standar deviasi sebagai parameter untuk menarik kesimpulan

In [ ]:
df.groupby('Outcome')[['Pregnancies', 'Glucose', 'BloodPressure',	'SkinThickness', 'Insulin',	'BMI',	'DiabetesPedigreeFunction', 'Age']].std()

Pregnancies    Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                     
0           3.017185  26.141200      18.063075      14.889947   98.865289   
1           3.741239  31.939622      21.491812      17.679711  138.689125   

              BMI  DiabetesPedigreeFunction        Age  
Outcome                                                 
0        7.689855                  0.299085  11.667655  
1        7.262967                  0.372354  10.968254

- Standar deviasi yang lebih tinggi dalam beberapa fitur pada kelompok yang memiliki diabetes (Outcome 1) menunjukkan bahwa variasi dalam fitur-fitur tersebut lebih besar dalam kelompok ini, menunjukkan tingkat heterogenitas yang lebih besar dalam data untuk kelompok ini.








---

### Identifikasi outlier dari kolom `Glucose` berdasarkan mean dan standar deviasi menggunakan `3-Sigma Rule`



In [ ]:
mean = df[['Glucose']].mean()
stdev = df[['Glucose']].std()
iso = (df[['Glucose']] < mean - 3*stdev) | (df[['Glucose']] > mean + 3*stdev)
df['is_outlier'] = iso

df1 = df[['Glucose', 'Outcome']].assign(is_outlier=iso)
df1.loc[df1['is_outlier']]

Glucose  Outcome  is_outlier
75         0        0        True
182        0        0        True
342        0        0        True
349        0        1        True
502        0        1        True

- Outlier pada Outcome 1: Ditemukan 2 outlier dalam kelompok yang memiliki diabetes. Hal ini menunjukkan adanya nilai-nilai yang sangat jauh dari sebagian besar data dalam kelompok ini. Outlier dalam kelompok yang memiliki diabetes dapat menunjukkan variasi yang ekstrem dalam beberapa fitur yang mungkin menjadi indikasi adanya kasus diabetes yang lebih parah atau tidak biasa.
- Outlier pada Outcome 0: Ditemukan 3 outlier dalam kelompok yang tidak memiliki diabetes. Hal ini juga menunjukkan adanya nilai-nilai yang sangat jauh dari sebagian besar data dalam kelompok ini. Outlier dalam kelompok yang tidak memiliki diabetes mungkin menunjukkan variasi ekstrem dalam karakteristik individu yang mungkin mempengaruhi risiko terhadap diabetes atau faktor-faktor lainnya yang relevan.

### Identifikasi extreme outlier dari kolom `Glucose` berdasarkan quartil pertama, ketiga, dan range antar kedua kuartil tersebut menggunakan *`Tukey's Fences`*



In [ ]:
q1 = df['Glucose'].quantile(q=0.25)
q3 = df['Glucose'].quantile(q=0.75)
iqr = q3 - q1
iso = (df['Glucose'] < q1 - 1.5*iqr) | (df['Glucose'] > q3 + 1.5*iqr)
iseo = (df['Glucose'] < q1 - 3*iqr) | (df['Glucose'] > q3 + 3*iqr)
df1 = df[['Glucose', 'Outcome']].assign(is_outlier = iso, is_extreme_outlier = iseo)
df1.loc[df1['is_outlier'] | df1['is_extreme_outlier']]

Glucose  Outcome  is_outlier  is_extreme_outlier
75         0        0        True               False
182        0        0        True               False
342        0        0        True               False
349        0        1        True               False
502        0        1        True               False

- Tidak ditemukan outlier yang extreme pada kedua outcome yang menunjukkan tidak ada nilai Glukosa yang secara signifikan jauh dari nilai-nilai lainnya dalam kelompok tersebut. Hal ini menunjukkan bahwa distribusi Glukosa mungkin relatif homogen atau tidak memiliki nilai ekstrem yang sangat jauh dari nilai-nilai lainnya.

## Correlation Analysis

In [ ]:
df.loc[:,'Pregnancies':].corr()

Pregnancies   Glucose  BloodPressure  SkinThickness  \
Pregnancies                  1.000000  0.129459       0.141282      -0.081672   
Glucose                      0.129459  1.000000       0.152590       0.057328   
BloodPressure                0.141282  0.152590       1.000000       0.207371   
SkinThickness               -0.081672  0.057328       0.207371       1.000000   
Insulin                     -0.073535  0.331357       0.088933       0.436783   
BMI                          0.017683  0.221071       0.281805       0.392573   
DiabetesPedigreeFunction    -0.033523  0.137337       0.041265       0.183928   
Age                          0.544341  0.263514       0.239528      -0.113970   
Outcome                      0.221898  0.466581       0.065068       0.074752   
is_outlier                  -0.025123 -0.306291      -0.006300       0.046024   

                           Insulin       BMI  DiabetesPedigreeFunction  \
Pregnancies              -0.073535  0.017683                 -0.033523   
Glucose                   0.331357  0.221071                  0.137337   
BloodPressure             0.088933  0.281805                  0.041265   
SkinThickness             0.436783  0.392573                  0.183928   
Insulin                   1.000000  0.197859                  0.185071   
BMI                       0.197859  1.000000                  0.140647   
DiabetesPedigreeFunction  0.185071  0.140647                  1.000000   
Age                      -0.042163  0.036242                  0.033561   
Outcome                   0.130548  0.292695                  0.173844   
is_outlier               -0.052857  0.009118                 -0.022413   

                               Age   Outcome  is_outlier  
Pregnancies               0.544341  0.221898   -0.025123  
Glucose                   0.263514  0.466581   -0.306291  
BloodPressure             0.239528  0.065068   -0.006300  
SkinThickness            -0.113970  0.074752    0.046024  
Insulin                  -0.042163  0.130548   -0.052857  
BMI                       0.036242  0.292695    0.009118  
DiabetesPedigreeFunction  0.033561  0.173844   -0.022413  
Age                       1.000000  0.238356   -0.031966  
Outcome                   0.238356  1.000000    0.008669  
is_outlier               -0.031966  0.008669    1.000000

- `Glucose` berkorelasi positif dengan `Outcome` menunjukkan bahwa ada hubungan yang cukup kuat antara Glukosa dan kemungkinan seseorang terkena diabetes (Outcome). Semakin tinggi tingkat glukosa dalam darah, semakin tinggi pula kemungkinan seseorang terkena diabetes. Glucose menjadi faktor yang paling berpengaruh dibanding faktor lain dalam dataset.
- Sementara itu, `BloodPressure` memiliki korelasi positif terendah dengan `Outcome` sehingga kita dapat menyimpulkan bahwa BloodPressure adalah faktor yang paling kecil mempengaruhi Outcome di antara faktor-faktor yang disediakan dalam data tersebut. Ini menunjukkan bahwa tidak ada hubungan yang kuat antara tekanan darah dan kemungkinan seseorang terkena diabetes berdasarkan data korelasi.

---